### PR TITLE
Fix memory error during precompilation

### LINKED ISSUE #44338: Pkg does not complete precompilation, randomly hangs on different packages
Since #43990 was merged with Julia master, I noticed that Pkg precompilation often hangs on some packages that I have installed, but there is no apparent pattern. It's also possible that the precompilation is just taking a really long time and appears to be hang. The first package I encountered this issue with was `Parsers.jl`. When this happens, ^C then repeat the precompilation usually completes the precompilation for the problematic packages. The circle animation spins indefinitely during the hanging episodes.

Here is one set:
```
Precompiling project...
  Progress [=======================================> ]  476/496
  ◑ StrideArraysCore
  ◐ MLJModels
^C Interrupted: Exiting precompilation...

(@ocean) pkg> update
    Updating registry at `~/.julia/registries/General.toml`
    Updating git-repo `https://github.com/Alexander-Barth/ROMS.jl`
  No Changes to `~/.julia/environments/ocean/Project.toml`
  No Changes to `~/.julia/environments/ocean/Manifest.toml`
Precompiling project...
  Progress [======================================>  ]  19/20
  ◐ MLJ
^C Interrupted: Exiting precompilation...
  18 dependencies successfully precompiled in 1098 seconds. 476 already precompiled. 1 skipped during auto due to previous errors.
  1 dependency errored. To see a full report either run `import Pkg; Pkg.precompile()` or load the package

julia> Pkg.precompile()
Precompiling project...
  2 dependencies successfully precompiled in 12 seconds. 494 already precompiled.
```

And another set:
```
Precompiling project...
  Progress [========================================>]  525/526
  ◓ IJulia
  ✗ JuliaFormatter
^C Interrupted: Exiting precompilation...

(@ocean) pkg> update
    Updating registry at `~/.julia/registries/General.toml`
  No Changes to `~/.julia/environments/ocean/Project.toml`
  No Changes to `~/.julia/environments/ocean/Manifest.toml`
Precompiling project...
  1 dependency successfully precompiled in 2 seconds. 524 already precompiled. 1 skipped during auto due to previous errors.
```

and another set:
```
(@ocean) pkg> update
    Updating registry at `~/.julia/registries/General.toml`
  No Changes to `~/.julia/environments/ocean/Project.toml`
  No Changes to `~/.julia/environments/ocean/Manifest.toml`
Precompiling project...
  Progress [========================================>]  515/526
  ◓ GeometryBasics
  ◑ PlotThemes
^C Interrupted: Exiting precompilation...

(@ocean) pkg> update
    Updating registry at `~/.julia/registries/General.toml`
  No Changes to `~/.julia/environments/ocean/Project.toml`
  No Changes to `~/.julia/environments/ocean/Manifest.toml`
Precompiling project...
  11 dependencies successfully precompiled in 28 seconds. 514 already precompiled. 1 skipped during auto due to previous errors.
```

versioninfo():

```
julia> versioninfo()
Julia Version 1.9.0-DEV.84
Commit 6b29ebda35 (2022-02-24 22:52 UTC)
Platform Info:
  OS: macOS (arm64-apple-darwin21.3.0)
  CPU: 10 × Apple M1 Pro
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-13.0.1 (ORCJIT, apple-m1)
  Threads: 1 on 8 virtual cores
```

### PR DESCRIPTION
**Pull request by timholy**
Fixes #44338

I tested this with
```
(pkgs) pkg> st
Status `/tmp/tim/pkgs/Project.toml`
  [5c1252a2] GeometryBasics v0.4.1
  [7073ff75] IJulia v1.23.2
  [98e50ef6] JuliaFormatter v0.22.4
  [add582a8] MLJ v0.17.1
  [d491faf4] MLJModels v0.15.4
  [ccf2f8ad] PlotThemes v2.0.1
  [7792a7ef] StrideArraysCore v0.2.12
```
which only intermittently triggers #44338, but the frequency was well above 1 in 10 and probably more like 1 in 3. After a few manual runs, I did this:

```sh
$ for i in {1..50}
> do
>     rm -rf ~/.julia/compiled/v1.9/*
>     ./julia --startup=no -e 'using Pkg; Pkg.activate("/tmp/tim/pkgs"); Pkg.precompile()'
>     echo "$i done"
> done
```

49 of these 50 runs completed successfully. The one exception threw this backtrace:

```
signal (11): Segmentation fault
in expression starting at none:0
jl_collect_backedges at /home/tim/src/julia-master/src/dump.c:1229
ijl_save_incremental at /home/tim/src/julia-master/src/dump.c:2676
jl_write_compiler_output at /home/tim/src/julia-master/src/precompile.c:65
ijl_atexit_hook at /home/tim/src/julia-master/src/init.c:207
jl_repl_entrypoint at /home/tim/src/julia-master/src/jlapi.c:707
jl_load_repl at /home/tim/src/julia-master/cli/loader_lib.c:271
main at /home/tim/src/julia-master/cli/loader_exe.c:59
__libc_start_main at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
_start at /home/tim/src/julia-master/usr/bin/julia-debug (unknown line)
Allocations: 8546957 (Pool: 8542808; Big: 4149); GC: 10
Stacktrace:
 [1] pkgerror(msg::String)
   @ Pkg.Types ~/src/julia-master/usr/share/julia/stdlib/v1.9/Pkg/src/Types.jl:67
 [2] precompile(ctx::Pkg.Types.Context, pkgs::Vector{String}; internal_call::Bool, strict::Bool, warn_loaded::Bool, already_instantiated::Bool, kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
   @ Pkg.API ~/src/julia-master/usr/share/julia/stdlib/v1.9/Pkg/src/API.jl:1427
 [3] precompile
   @ ~/src/julia-master/usr/share/julia/stdlib/v1.9/Pkg/src/API.jl:1060 [inlined]
 [4] #precompile#225
   @ ~/src/julia-master/usr/share/julia/stdlib/v1.9/Pkg/src/API.jl:1057 [inlined]
 [5] precompile (repeats 2 times)
   @ ~/src/julia-master/usr/share/julia/stdlib/v1.9/Pkg/src/API.jl:1057 [inlined]
 [6] top-level scope
   @ none:1
```

which is also htable-related, but one that we've used this way for a long time. In particular, the crash probably occurred in a part of that line that has been part of Julia for 5 years: https://github.com/JuliaLang/julia/blob/fb85cb3a5deea742260c70a2763aa7c64d26eb02/src/dump.c#L1077

One other interesting case: while almost all of these runs took between 90-92s each to compile, on one run it took 436s. I don't have any data about what caused that, so it's not really actionable.

### CODE REVIEW

**Comment by truedichotomy** [comment]
Thanks for acting on this bug so quickly!

**Review by vtjnash — DISMISSED** [review]
This cannot be correct, since it doesn't change the access patterns, just the initialization size

**Comment by timholy** [comment]
Operationally it fixes the problem, though I agree it's not the cause. Just checking, you saw https://github.com/JuliaLang/julia/issues/44338#issuecomment-1050869978, right? The hang happens when the `htable` resizes, so I figured preventing the need to resize is not an unreasonable way to fix this.

**Comment by JeffBezanson** [comment]
Looks like memory corruption? E.g. some of malloc's internal state getting clobbered.

**Comment by timholy** [comment]
I'll try firing up `valgrind`. FWIW `analyzegc` comes clean on #43990 (if I'm interpreting its output correctly).

**Comment by truedichotomy** [comment]
Tested this PR and it works with my system & all the packages.

**Comment by timholy** [comment]
Valgrind suggests there's a leak:
```
==2151518== 13,312 bytes in 8 blocks are definitely lost in loss record 11,916 of 12,166
==2151518==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==2151518==    by 0x51D167E: equalhash_lookup_bp_r (equalhash.c:22)
==2151518==    by 0x51D16EF: equalhash_put_r (equalhash.c:22)
==2151518==    by 0x51D2202: fl_table_put (table.c:116)
==2151518==    by 0x51C37CD: apply_cl (flisp.c:1230)
==2151518==    by 0x51C7540: _applyn (flisp.c:685)
==2151518==    by 0x51C7D9C: fl_map1 (flisp.c:2223)
==2151518==    by 0x51C7D9C: fl_map1 (flisp.c:2198)
==2151518==    by 0x51C37CD: apply_cl (flisp.c:1230)
==2151518==    by 0x51C7540: _applyn (flisp.c:685)
==2151518==    by 0x51C7D9C: fl_map1 (flisp.c:2223)
==2151518==    by 0x51C7D9C: fl_map1 (flisp.c:2198)
==2151518==    by 0x51C37CD: apply_cl (flisp.c:1230)
==2151518==    by 0x51C8114: do_trycatch (flisp.c:901)
```
There are several "copies" of this report. But of all the htable uses the only leak seems to come from here.

**Comment by JeffBezanson** [comment]
Definitely looks like a job for rr.

**Comment by timholy** [comment]
Found it! Phew.

There's no reason I can see not to do the pre-allocation, so I decided to add the real fix to this PR. But I've tested this change in isolation, and it solves the problem on its own (>1000 successful precompilation of Crayons without a hitch).

**Comment by DilumAluthge** [comment]
@vtjnash Does this look good to you now?

**Comment by timholy** [comment]
This definitively fixes a serious bug and there is otherwise not much to this. I don't think it needs to wait on a review.

File: src/dump.c [hunk 1]
```diff
@@ -286,10 +287,10 @@ static size_t queue_external_mis(jl_array_t *list)
 {
     size_t i, n = 0;
     htable_t visited;
-    htable_new(&visited, 0);
     if (list) {
         assert(jl_is_array(list));
         size_t n0 = jl_array_len(list);
+        htable_new(&visited, n0);
```
Comment (vtjnash) [inline src/dump.c:293]: The `htable_free` below now access uninitialized memory and must be moved also

**Comment by vtjnash** [comment]
I don't know this part of the code, but the change makes sense now

File: src/dump.c (hunk 1 above)
Comment (timholy) [reply src/dump.c:293]: Thanks! Fixed in #44446

**Pull request merged**